### PR TITLE
fix(allocator): fixed-size allocators use `VirtualAlloc` on Windows

### LINKED ISSUE #19395: linter: panic: Insufficient memory to create fixed-size allocator pool
### What version of Oxlint are you using?

1.47.0

### What command did you run?

oxlint --fix

### What does your `.oxlintrc.json` config file look like?

```jsonc
{
  "plugins": ["eslint", "import", "jsdoc", "jsx-a11y", "unicorn", "oxc", "typescript", "promise"],
  "categories": {
    "correctness": "error",
    "perf": "error"
  },
  "overrides": [
    {
      "files": ["src/**/*.{ts,tsx}"],
      "rules": {
        "typescript/explicit-function-return-type": "error"
      }
    },
    {
      "files": [
        "src/**/*.test.{ts,tsx}",
        "src/**/*.spec.{ts,tsx}",
        "src/**/__tests__/**/*.{ts,tsx}"
      ],
      "rules": {
        "typescript/explicit-function-return-type": "off"
      }
    },
    {
      "files": ["src/**/*.tsx", "playground/**/*.tsx"],
      "jsPlugins": ["eslint-plugin-solid", "@unocss/eslint-plugin"],
      "rules": {
        "@unocss/order": "warn",
        "solid/event-handlers": [
          "error",
          {
            // if true, don't warn on ambiguously named event handlers like `onclick` or `onchange`
            "ignoreCase": false,
            // if true, warn when spreading event handlers onto JSX. Enable for Solid < v1.6.
            "warnOnSpread": false
          }
        ],
        // these rules are mostly style suggestions
        "solid/imports": "error",
        // identifier usage is important
        "solid/jsx-no-duplicate-props": "error",
        "solid/jsx-no-script-url": "error",
        "solid/no-destructure": "error",
        // security problems
        "solid/no-innerhtml": ["error", { "allowStatic": true }],
        "solid/no-react-deps": "error",
        "solid/no-react-specific-props": "error",
        "solid/no-unknown-namespaces": "error",
        "solid/prefer-for": "error",
        "solid/prefer-show": "error",
        "solid/self-closing-comp": "error",
        "solid/style-prop": ["error", { "styleProps": ["style", "css"] }],
        "solid/jsx-no-undef": ["error", { "typescriptEnabled": true }],
        "solid/reactivity": "warn"
      }
    }
  ],
  "rules": {
    "curly": ["error", "all"],
    "func-style": [
      "error",
      "declaration",
      {
        "allowArrowFunctions": true
      }
    ],
    "default-case-last": "error",
    "default-param-last": "error",
    "no-await-in-loop": "allow",
    "promise/no-callback-in-promise": "allow",
    "array-callback-return": "error",
    "eqeqeq": ["error", "smart"],
    "getter-return": "error",
    "import/named": "error",
    "for-direction": "error",
    "require-yield": "error",
    "use-isnan": "error",
    "valid-typeof": "error",
    "no-empty-character-class": "error",
    "no-self-compare": "error",
    "no-lone-blocks": "error",
    "no-empty": [
      "error",
      {
        "allowEmptyCatch": true
      }
    ],
    "no-constant-binary-expression": "error",
    "no-constructor-return": "error",
    "no-func-assign": "error",
    "no-global-assign": "error",
    "no-unused-vars": "error",
    "no-ex-assign": "error",
    "no-loss-of-precision": "error",
    "no-fallthrough": "error",
    "no-nonoctal-decimal-escape": "error",
    "no-irregular-whitespace": "error",
    "no-control-regex": "error",
    "no-with": "error",
    "no-debugger": "error",
    "no-const-assign": "error",
    "no-setter-return": "error",
    "no-obj-calls": "error",
    "no-unsafe-negation": "error",
    "no-dupe-class-members": "error",
    "no-dupe-keys": "error",
    "no-this-before-super": "error",
    "no-useless-catch": "error",
    "no-async-promise-executor": "error",
    "no-unreachable": "error",
    "no-duplicate-case": "error",
    "no-empty-pattern": "error",
    "no-unused-labels": "error",
    "no-sparse-arrays": "error",
    "no-delete-var": "error",
    "no-compare-neg-zero": "error",
    "no-case-declarations": "error",
    "no-class-assign": "error",
    "no-var": "error",
    "no-self-assign": "error",
    "no-inner-declarations": "error",
    "no-dupe-else-if": "error",
    "no-invalid-regexp": "error",
    "no-unsafe-finally": "error",
    "no-prototype-builtins": "error",
    "no-array-constructor": "error",
    "no-shadow-restricted-names": "error",
    "no-constant-condition": "error",
    "no-useless-escape": "error",
    "no-unsafe-optional-chaining": "error",
    "no-extra-boolean-cast": "error",
    "no-regex-spaces": "error",
    "typescript/consistent-type-definitions": "error",
    "typescript/consistent-type-imports": "error",
    "typescript/no-this-alias": [
      "error",
      {
        "allowDestructuring": true
      }
    ],
    "typescript/prefer-as-const": "error",
    "typescript/no-var-requires": "error",
    "typescript/prefer-ts-expect-error": "error",
    "typescript/ban-ts-comment": [
      "error",
      {
        "ts-expect-error": "allow-with-description",
        "ts-ignore": false,
        "ts-nocheck": false,
        "ts-check": false
      }
    ],
    "unicorn/no-useless-promise-resolve-reject": "error",
    "unicorn/no-unnecessary-await": "error",
    "unicorn/no-useless-fallback-in-spread": "error",
    // "unicorn/prefer-dom-node-dataset": "error",
    // "unicorn/prefer-dom-node-append": "error",
    // "unicorn/prefer-dom-node-remove": "error",
    "unicorn/prefer-date-now": "error",
    "unicorn/prefer-blob-reading-methods": "error",
    "unicorn/no-typeof-undefined": "error",
    "unicorn/no-new-array": "error",
    "unicorn/new-for-builtins": "error",
    "unicorn/prefer-node-protocol": "error",
    "import/no-cycle": [
      "error",
      {
        "ignoreTypes": true
      }
    ],
    "import/consistent-type-specifier-style": "error",
    "import/sort-imports": "off",
    "import/no-duplicates": "error",
    "import/no-import-assign": "error",
    "import/no-self-import": "error",
    "import/no-named-export": "allow",
    "jsx-a11y/label-has-associated-control": "off",
    "jsx-a11y/prefer-tag-over-role": "off"
  }
}

```

### What happened?

On Windows 11 , sometimes it will run with error, but after toggling `jsplugin`, the error has gone.

```
thread 'tokio-runtime-worker' (7992) panicked at crates\oxc_allocator\src\pool\fixed_size.rs:170:18:
Insufficient memory to create fixed-size allocator pool
stack backtrace:
   0:     0x7ffd12c86f82 - napi_register_module_v1
   1:     0x7ffd12c92db1 - napi_register_module_v1
   2:     0x7ffd12c68614 - napi_register_module_v1
   3:     0x7ffd12c7cc70 - napi_register_module_v1
   4:     0x7ffd12c7ca88 - napi_register_module_v1
   5:     0x7ffd12c7cfa2 - napi_register_module_v1
   6:     0x7ffd12c7cd79 - napi_register_module_v1
   7:     0x7ffd12c77c2f - napi_register_module_v1
   8:     0x7ffd12c653de - napi_register_module_v1
   9:     0x7ffd12d10fbd - napi_register_module_v1
  10:     0x7ffd12c459bd - napi_register_module_v1
  11:     0x7ffd127eb231 - napi_register_module_v1
  12:     0x7ffd12711cec - napi_register_module_v1
  13:     0x7ffd1258591a - <unknown>
  14:     0x7ffd12541bc3 - <unknown>
  15:     0x7ffd125546c8 - <unknown>
  16:     0x7ffd1268a02b - napi_register_module_v1
  17:     0x7ffd1268825b - napi_register_module_v1
  18:     0x7ffd12682390 - napi_register_module_v1
  19:     0x7ffd1267e3dd - napi_register_module_v1
  20:     0x7ffd1267ebb1 - napi_register_module_v1
  21:     0x7ffd12c70cc8 - napi_register_module_v1
  22:     0x7ffdb2a4259d - BaseThreadInitThunk
  23:     0x7ffdb2f6af38 - RtlUserThreadStart
error: script "qa" exited with code 9
```

My current workaround
```json
{
  "scripts": {
    "fix-oxlint": "bun -e \"const p=process.argv[1],f=Bun.file(p),t=await f.text();await Bun.write(p,t.replace(/jsPlugins/g,'//jsPlugins'));await Bun.write(p,t)\" .oxlintrc.json",
    "oxc": "bun run fix-oxlint && oxlint --fix && oxfmt"
  }
}
```

### PR DESCRIPTION
**Pull request by overlookmotel**
Fixes #19395 - OOM errors in Oxlint JS plugins on Windows.

### The problem

Previously, on all platforms each fixed-size (raw transfer-enabled) allocator obtained a large chunk of memory from the `System` allocator (6 GiB in the case of Windows). This is fine on MacOS and most Linux distros, as they use overcommit allocation accounting, so these allocations consume only virtual memory (address space), not actual physical memory. The available address space is enormous (~128 TiB), so it's hard to exhaust it. Physical memory is only consumed when pages of the allocation are actually touched, which the vast majority never are.

Windows behaves differently. Allocations obtained from `System` allocator immediately count towards the memory limit. So, on memory-constrained machines, allocating 6 GiB can immediately fail - OOM just from linting the first file.

### This PR

[#22088](https://github.com/oxc-project/oxc/issues/22088), #22122, and #22123 introduced platform-specific code for handling allocation and deallocation of fixed-size `Arena`s, which are used in raw transfer/Oxlint JS plugins.

This PR switches to a different allocation method on Windows for fixed-size `Arena`s.

It switches to using [VirtualAlloc](https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc). This Windows-only API allows reserving address space and committing pages of that reservation as separate operations (similar to `mmap`).

It works like this:

- Fixed-size allocators now still reserve 6 GiB, but that only consumes address space, and does not count towards max memory limit.
- Of that 6 GiB reservation, a 2 GiB "container", aligned on a 4 GiB boundary, will be used as the active part of the memory for storing the AST and other data.
- Initially only the last 16 KiB of the container is committed (made into usable memory, with physical pages backing it).
- If AST does not fit in the committed region, the region grows downwards, committing further pages to accomodate allocation requests, up to the maximum of 2 GiB.

Essentially, this replicates something similar to Linux-style allocation, but on Windows. The only difference is that on Windows it's necessary to explicitly commit pages before touching them, whereas Linux/MacOS commit on demand automatically.

### Implementation details and testing

This PR uses the battle-tested [windows-sys](https://crates.io/crates/windows-sys) crate for `VirtualAlloc` and `VirtualFree`. The allocation implementation is written from scratch, but while I was researching to validate that I'd not missed any edge cases, I discovered that it's almost identical to [wasmtime's implementation](https://github.com/bytecodealliance/wasmtime/blob/386a3280dee61f5c4120ce7cde621c1039e383d5/crates/wasmtime/src/runtime/vm/sys/windows/mmap.rs) which uses the same APIs. Comments in the code go over in some detail considerations of Rust's memory model and pointer provenance. As far as I can see, the implementation is fairly watertight.

There are tests for the essential functionality and edge cases, and #22121 made these tests run under Miri on Windows.

However, I don't have access to a Windows machine to fully test this by running Oxlint on a massive repo. I am hoping members of the community who do can put this through its paces a bit before we merge.

### What's missing

In enormous projects, when linting with both JS plugins and `import` plugin, it's possible to end up with a huge number of ASTs in memory simultaneously. With each AST consuming 6 GiB of address space, it's possible to exhaust the entire 128 TiB usable address space, and get OOM on _virtual_ memory.

At present we have a workaround to prevent too many fixed-size allocators existing concurrently, so avoiding OOM, but the workaround is a drag on performance. Future work will aim to address this.

### CODE REVIEW

**Comment by overlookmotel** [comment]
* **#22424** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/22424?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22392** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/22392?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22124** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/22124?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/22124?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#22123** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/22123?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22122** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/22122?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22121** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/22121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22388** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/22388?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent changes, fast-track this PR to the front of the merge queue

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=22124).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/oxc_allocator/src/arena/fixed_size/windows.rs [hunk 1]
```diff
@@ -117,27 +338,111 @@ impl<const MIN_ALIGN: usize> Arena<MIN_ALIGN> {
...
+        //    at start of Container. Same logic as the fully-committed case. We must not `VirtualAlloc(MEM_COMMIT)`
+        //    on these - the backing memory wasn't reserved via `VirtualAlloc`.
+        //
+        // Note: `napi/parser` creates an `Arena` with `Arena::from_raw_parts`, so with `is_fixed_size: true`.
+        // But it *doesn't* have `start_ptr` aligned on `CONTAINER_ALIGN`. In extreme cases, this could allow this check
+        // to pass when it shouldn't, and attempt to commit memory which wasn't reserved via `VirtualAlloc`.
+        // However, `napi/parser` doesn't enable `fixed_size` Cargo feature, so this code is not compiled,
+        // and therefore there's no way for this failure to arise.
+        // TODO: This is fragile. Fix it by aligning `start_ptr` to `CONTAINER_ALIGN` in `napi/parser`
+        // (write source text into end of buffer, not start).
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_allocator/src/arena/fixed_size/windows.rs:380]: This is exactly what the comment here says. It's not possible for that to occur, but it's fragile. I'll harden it up shortly. At present, there's is no way to make this more robust without adding a ton of code which will all be removed again in a few days anyway.

File: crates/oxc_allocator/src/arena/fixed_size/windows.rs [hunk 2]
```diff
@@ -117,27 +338,111 @@ impl<const MIN_ALIGN: usize> Arena<MIN_ALIGN> {
...
 ///
+/// This releases the entire reservation (committed and uncommitted parts) via `VirtualFree`.
+///
 /// # SAFETY
 ///
 /// * `footer_ptr` must point to a valid `ChunkFooter`.
-/// * `ChunkFooter` must be for a fixed size chunk (created via `Arena::from_raw_parts` or `Arena::new_fixed_size`).
+/// * `ChunkFooter` must be for a fixed size chunk.
+/// * Chunk must have been created via `Arena::new_fixed_size`.
 pub unsafe fn dealloc_fixed_size_arena_chunk(footer_ptr: NonNull<ChunkFooter>) {
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_allocator/src/arena/fixed_size/windows.rs:442]: See previous comment. There is some ambiguity and mess here. It's definitely fragile, but there is no code path which can commit UB. I'll harden it up in next week.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -311,39 +311,40 @@ jobs:
...
+        #if: steps.filter.outputs.changed == 'true'
         with:
           babel: false
           prettier: false
           node-compat-table: false
-      - if: steps.filter.outputs.changed == 'true'
-        name: Run tests
+      - name: Run tests
+        #if: steps.filter.outputs.changed == 'true'
         working-directory: ${{ env.DEV_DRIVE_WORKSPACE }}
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply .github/workflows/ci.yml:348]: That's a temp commit, now reverted. Just wanted CI to run on Windows to check it's green.

**Comment by Sysix** [comment]
@overlookmotel 
I tested it today with a big project, where the current oxlint panicked after 5 minutes of opening and modifying heavy `jsPlugins` files.
I tried to be as much RAM consuming, as possible. Started a 10+ docker container setup, file watcher, HD videos and more.

With this version, I could not produce a panic after 20 minutes. Almost removed jQuery from an 800 line file. Modified other heavy files too, and it worked like a charm.

I can test it on Monday with a longer session. But it looks like it fixes the issue in windows for me 👍

**Comment by overlookmotel** [comment]
> I tried to be as much RAM consuming, as possible. Started a 10+ docker container setup, file watcher, HD videos and more.

Hahaha brilliant! You really went to town! Would love to have seen your machine rocking under *that* load. "HD videos and more" is the crowning touch!

Thanks loads for putting it through its paces.

I got cold feet about merging it before today's release. I also want to put it through its paces on Linux/Mac to make sure the changes don't accidentally "bleed" into other platforms. Please excuse my excess caution - destabilizing the allocator could destabilize everything. I intend to look at it again later this week, do more testing, and hopefully get it into next Monday's release.

**Comment by Diggsey** [comment]
This has taken the "Private Bytes" for oxlint from 88GB to ~100MB, a 99.9% reduction in committed memory 🎉 .

**Comment by Sysix** [comment]
@overlookmotel could you sync the graphite stack? 
I would like to retry it. I could only check on Monday with a small 2~ hour session. Next days I will have more frontend tasks, where `oxlint --lsp` is relevant. Hope for a 4+ hour session :)

**Comment by overlookmotel** [comment]
Stack now rebased on main.

However, I'm going to do a fix for #22339 tomorrow, and then will have to alter this stack too to go on top of that fix.

**Comment by Sysix** [comment]
Could not crash the LSP server today 🥳

File: crates/oxc_allocator/src/arena/alloc_impl.rs [hunk 4]
```diff
@@ -335,7 +335,22 @@ impl<const MIN_ALIGN: usize> Arena<MIN_ALIGN> {
...
+                // `fixed_size` feature is only enabled in Oxlint.
+                // * In Oxlint itself, `grow_fixed_size_chunk` will attempt to grow the chunk in place.
+                // * In Oxlint's `RuleTester`, `Arena`s have `is_fixed_size: true`, but `start_ptr` is aligned on 4 GiB,
+                //   so `grow_fixed_size_chunk` will always fail to grow the chunk, and will return `None`.
+                // * In `napi/parser` with raw transfer enabled, `Arena`s also have `is_fixed_size: true`.
+                //   `start_ptr` is *not* aligned on 4 GiB, so calling `grow_fixed_size_chunk` would be UB.
+                //   (on Windows it would attempt to call `VirtualAlloc` with a pointer which was part of an allocation
+                //   made by JS runtime, not `VirtualAlloc`).
+                //   But `napi/parser` does not enable `fixed_size` Cargo feature, so this whole block is skipped,
+                //   and this function return `None`.
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_allocator/src/arena/alloc_impl.rs]: Fixed.

File: crates/oxc_allocator/src/from_raw_parts.rs [hunk 5]
```diff
@@ -29,8 +29,13 @@ impl Allocator {
...
     ///
     /// The [`Allocator`] which is returned takes ownership of the backing allocation, and it will be freed
-    /// via the [`System`] allocator (using `backing_alloc_ptr` and `layout`) when the `Allocator` is dropped.
-    /// If caller wishes to prevent that happening, they must wrap the `Allocator` in `ManuallyDrop`.
+    /// when the `Allocator` is dropped, using `backing_alloc_ptr` and `layout`.
+    ///
+    /// The method used to free the backing allocation is platform-dependent:
+    /// * Linux/Mac: via [`System`] allocator
+    /// * Windows: `VirtualFree`
+    ///
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_allocator/src/from_raw_parts.rs]: True. Have updated the comment.

File: crates/oxc_allocator/src/arena/from_raw_parts.rs [hunk 6]
```diff
@@ -19,8 +19,13 @@ impl<const MIN_ALIGN: usize> Arena<MIN_ALIGN> {
...
     ///
     /// The [`Arena`] which is returned takes ownership of the backing allocation, and it will be freed
-    /// via the [`System`] allocator (using `backing_alloc_ptr` and `layout`) when the `Arena` is dropped.
-    /// If caller wishes to prevent that happening, they must wrap the `Arena` in `ManuallyDrop`.
+    /// when the `Arena` is dropped, using `backing_alloc_ptr` and `layout`.
+    ///
+    /// The method used to free the backing allocation is platform-dependent:
+    /// * Linux/Mac: via [`System`] allocator
+    /// * Windows: `VirtualFree`
+    ///
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_allocator/src/arena/from_raw_parts.rs]: True. Have updated the comment.

File: crates/oxc_allocator/src/arena/alloc_impl.rs [hunk 7]
```diff
@@ -335,7 +335,22 @@ impl<const MIN_ALIGN: usize> Arena<MIN_ALIGN> {
             // SAFETY: `footer_ptr` always points to a valid `ChunkFooter`
             let footer = unsafe { footer_ptr.as_ref() };
             if footer.is_fixed_size {
-                // Attempt to grow the chunk in place to accommodate the allocation
+                // Attempt to grow the chunk in place to accommodate the allocation.
+                //
+                // `fixed_size` feature is only enabled in Oxlint.
+                // * In Oxlint itself, `grow_fixed_size_chunk` will attempt to grow the chunk in place.
+                // * In Oxlint's `RuleTester`, `Arena`s have `is_fixed_size: true`, but `start_ptr` is aligned on 4 GiB,
+                //   so `grow_fixed_size_chunk` will always fail to grow the chunk, and will return `None`.
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_allocator/src/arena/alloc_impl.rs:344]: I kind of hate, kind of love what an absolute pedant you are. Fixed.

File: crates/oxc_allocator/Cargo.toml [hunk 8]
```diff
@@ -38,6 +38,9 @@ rustc-hash = { workspace = true }
 
 serde = { workspace = true, optional = true }
 
+[target.'cfg(target_os = "windows")'.dependencies]
+windows-sys = { workspace = true, features = ["Win32_System_Memory"] }
```
[the remark this answers is not in the document]
Comment (overlookmotel) [reply crates/oxc_allocator/Cargo.toml]: Yes, you're right. One less dependency in compiler pipeline on Windows. Fixed.

**Comment by overlookmotel** [comment]
Have tested this branch on Mac OS linting Kibana repo with JS plugins + `import` plugin combined. We've found this the most demanding repo to test Oxlint on in past. It works as expected, so it looks like this change isn't causing problems on other platforms.

Finally going to merge this thing!
